### PR TITLE
feat: Process orders and combinations

### PR DESCRIPTION
**Pull request by maximebrugel**
We are now introducing Batched**Input**Orders and Batched**Output**Orders, to extend the possibilities beyond the multi-swap (without allowing everything) and simplify the interface.

#### Batched Input Orders
image
- One same input for every orders but multiple outputs.
- 1% Fee on the input.
- The input (*source*) is from a wallet **or** a portfolio.
- The ouput (*destination*) is the portfolio (**only**). 

```javascript
struct BatchedInputOrders {
    IERC20 inputToken;
    uint256 amount;
    Order[] orders;
    bool fromReserve;
}
```

#### Batched Output Orders
image
- Multiple inputs for every orders but one output.
- 1% Fee on the output.
- The input (*source*) is the portfolio (**only**). 
- The output (*destination*) is from a wallet **or** a portfolio.

```javascript
struct BatchedOutputOrders {
    IERC20 outputToken;
    uint256[] amounts;
    Order[] orders;
    bool toReserve;
}
```

We removed multiple functions: 
- `addTokens`
- `swapTokenForTokens `
- `swapTokensForTokens`
- `sellTokensToNft`

And added new ones : 

```javascript
/// @notice Create a portfolio and store the underlying assets from the positions
/// @param _originalTokenId The id of the NFT replicated, 0 if not replicating
/// @param _batchedOrders The order to execute
function create(uint256 _originalTokenId, BatchedInputOrders[] calldata _batchedOrders) external payable;

/// @notice Process multiple input orders
/// @param _nftId The id of the NFT to update
/// @param _batchedOrders The order to execute
function processInputOrders(uint256 _nftId, BatchedInputOrders[] calldata _batchedOrders) external payable;

/// @notice Process multiple output orders
/// @param _nftId The id of the NFT to update
/// @param _batchedOrders The order to execute
function processOutputOrders(uint256 _nftId, BatchedOutputOrders[] calldata _batchedOrders) external;

/// @notice Process multiple input orders and then multiple output orders
/// @param _nftId The id of the NFT to update
/// @param _batchedInputOrders The input orders to execute (first)
/// @param _batchedOutputOrders The output orders to execute (after)
function processInputAndOutputOrders(
    uint256 _nftId,
    BatchedInputOrders[] calldata _batchedInputOrders,
    BatchedOutputOrders[] calldata _batchedOutputOrders
) external payable;
```

_Note : The `create` function is not new, but is now allowing to make multiple `BatchedInputOrders`._

### New transfer input tokens ⚠️

Now that we can make multiple `BatchedInputOrders`, we can use ETH as `inputToken` multiple times.
The `_transferInputTokens` function was using `msg.value` to `deposit` in the WETH contract and check if the `msg.value` amount was valid. Therefore, this way of handling ETH will fail if we make more than one `BatchedInputOrders`.

To handle this case : 
```javascript
if (address(_inputToken) == ETH) {
    require(address(this).balance >= _inputTokenAmount, "NF: INVALID_AMOUNT_IN");
    weth.deposit{ value: _inputTokenAmount }();
    return (IERC20(address(weth)), _inputTokenAmount);
}
```

The function will check if the balance of ETH is enough to deposit the `_inputTokenAmount` and will `deposit` this amount. We are consuming the ETH balance for each input orders.
But there is a downside, a user can send more ETH than the sum of `BatchedInputOrders.amount` (in case of an error in the front-end implementation). And someone could create/edit a portfolio with the `NestedFactory` Ethers , because we are not checking `msg.value`.

To avoid this issue, we need to check the `msg.value` with this function:
```javascript
/// @dev Verify that msg.value is equal to the amount needed (in the orders)
/// @param _batchedOrders The batched input orders
function _checkMsgValue(BatchedInputOrders[] calldata _batchedOrders) private {
    uint256 ethNeeded;
    uint256 length = _batchedOrders.length;
    for (uint256 i = 0; i < length; i++) {
        if (address(_batchedOrders[i].inputToken) == ETH) {
            ethNeeded += _batchedOrders[i].amount;
        }
    }
    require(msg.value == ethNeeded, "NF: WRONG_MSG_VALUE"); 
}
```
This function is called at the beginning of every payable functions, to check `msg.value`.

### CODE REVIEW

File: contracts/NestedFactory.sol [hunk 1]
```diff
@@ -271,34 +229,62 @@ contract NestedFactory is INestedFactory, ReentrancyGuard, Ownable, MixinOperato
...
+    /// @param _toReserve True if the output is store in the reserve/records, false if not.
     /// @param _fromReserve True if the input tokens are from the reserve
     /// @return feesAmount The total amount of fees
     /// @return tokenSold The ERC20 token sold (in case of ETH to WETH)
     function _submitInOrders(
         uint256 _nftId,
-        BatchedOrders calldata _batchedOrders,
-        bool _reserved,
+        BatchedInputOrders calldata _batchedOrders,
+        bool _toReserve,
```
Comment (adrien-supizet) [inline contracts/NestedFactory.sol:287]: ❔ Isn't `_toReserve` always true for a buy order?

File: README.MD [hunk 2]
```diff
@@ -57,8 +57,8 @@ They kind of work like [libraries](https://docs.soliditylang.org/en/v0.8.9/contr
 they are contracts that are called via `delegatecall` and referenced by the `OperatorResolver`.
 
 #### One language
-To interact with new operators on the fly, they must speak the same language.
-Let's define the "interactions" common to all operators: `commit` and `revert`.
+To interact with new operatrs on the fly, they must speak the same language.
+Let's define the functions" common to all operators: `commit` and `revert`.
```
[the remark this answers is not in the document]
Comment (maximebrugel) [reply README.MD:61]: The README is updated in the operator simplifications PR => https://github.com/NestedFinance/nested-core-lego/pull/97/commits/142e68b288ee7a647e3b20e1434750b093a44c9f

File: README.MD [hunk 3]
```diff
@@ -97,6 +97,61 @@ This way, the operator can retrieve the storage address (without storing it) and
 
 _More operators will be added. e.g. CurveOperator or SynthetixOperator_
 
+### Orders
+
+The `NestedFactory` is using the operators to interact with other protocols. The call from the Factory to an Operator is an "Order".
+
+An Order has several information:
+- The operator to use.
+- If it should use "Commit" or "Revert".
```
[the remark this answers is not in the document]
Comment (maximebrugel) [reply README.MD:106]: I will update the documentation after merging the two PRs

File: contracts/NestedFactory.sol (hunk 1 above)
Comment (maximebrugel) [reply contracts/NestedFactory.sol:289]: Yes! This PR will remove useless params => https://github.com/NestedFinance/nested-core-lego/pull/68

File: contracts/NestedFactory.sol [hunk 4]
```diff
@@ -558,4 +561,17 @@ contract NestedFactory is INestedFactory, ReentrancyGuard, Ownable, MixinOperato
         _transferFeeWithRoyalty(feeAmount, _token, _nftId);
         _token.safeTransfer(_dest, _amount - feeAmount);
     }
+
+    /// @dev Verify that msg.value is equal to the amount needed (in the orders)
+    /// @param _batchedOrders The batched input orders
+    function _checkMsgValue(BatchedInputOrders[] calldata _batchedOrders) private {
+        uint256 ethNeeded;
+        uint256 length = _batchedOrders.length;
```
Comment (adrien-supizet) [inline contracts/NestedFactory.sol:569]: 💅 the length caching is not required here

**Review by adrien-supizet — COMMENTED** [review]
Looks good to me! 👏

File: contracts/NestedFactory.sol (hunk 4 above)
Comment (maximebrugel) [reply contracts/NestedFactory.sol]: Changed 👍

**Pull request merged**